### PR TITLE
Allow custom effects to customize the inspector displayed in the material

### PR DESCRIPTION
**Pull request by chichinohaha**
Re: cocos-creator/3d-tasks#

Changelog:
 * Now we can customize the presentation of the effect in the material's inspector like this.
  image
Suppose we have an enabled extension named 'ss'
and there is a hack.js file in that plugin directory
This file is similar to the inspector,
but will get the currently used material in the first argument of the update method
example:
```js
// ss/hack.js
"use strict";
module.exports = Editor.Panel.define({
    template:
    /* html*/ `
    <ui-prop>
        <ui-label slot="label"></ui-label>
        <ui-checkbox slot="content"></ui-checkbox>
    </ui-prop>
    `,
    $: {
        checkbox: 'ui-checkbox',
        label: 'ui-label',
    },
    ready() {
        var _a;
        (_a = this.$.checkbox) === null || _a === void 0 ? void 0 : _a.addEventListener('change', (event) => {
            this.material.data[0].passes[0].defines[0].value = event.target.value;
            this.dispatch('change');
        });
    },
    update(material, assetList, metaList) {
        this.material = material;
        this.$.label.value = this.material.data[0].passes[0].defines[0].name;
        this.$.checkbox.value = this.material.data[0].passes[0].defines[0].value;
    },
});

```
 
image

### CODE REVIEW

File: editor/i18n/zh/assets.js [hunk 1]
```diff
@@ -298,5 +298,8 @@ module.exports = {
             modeWarn:
             '警告：WebGL 1.0 平台不支持非 2 次幂贴图的 repeat 过滤模式，运行时会自动改为 clamp-to-edge 模式，这会使材质的 tilingOffset 等属性完全失效。',
         },
+        material: {
+            'fail-to-load-custom-inspector': 'material: 自定义材质 {effect} 的 inspector 加载失败',
```
Comment (arsen2010) [inline editor/i18n/zh/assets.js]: “自定义材质” change to “自定义 Effect”

File: editor/inspector/assets/material.js [hunk 2]
```diff
@@ -40,9 +44,19 @@ exports.$ = {
     materialDump: '.material-dump',
     useInstancing: '.useInstancing',
     useBatching: '.useBatching',
+    customSection: '.customSection',
+    customPanel: '.customPanel',
 };
 
 exports.methods = {
+    async getCustomInspector() {
```
Comment (arsen2010) [inline editor/inspector/assets/material.js:52]: advise use 'hasCustomInspector' as function name

File: editor/inspector/assets/material.js [hunk 3]
```diff
@@ -40,9 +44,19 @@ exports.$ = {
...
+        const currentEffectInfo = this._effects.find(effect => { return effect.name === this.material.effect; });
+        if (currentEffectInfo && currentEffectInfo.uuid) {
+            const meta = await Editor.Message.request('asset-db', 'query-asset-meta', currentEffectInfo.uuid);
+            return meta && meta.userData && meta.userData.editor && meta.userData.editor.inspector;
```
Comment (arsen2010) [inline editor/inspector/assets/material.js:56]: return true or false

File: editor/inspector/assets/material.js [hunk 4]
```diff
@@ -56,6 +70,25 @@ exports.methods = {
         this.dirtyData.uuid = '';
     },
 
+    async updateCustomInspector() {
+        this.$.customPanel.hidden = false;
+        this.$.section.hidden = true;
+        this.$.materialDump.hidden = true;
+        try {
+            const relatePath = normalize((await this.getCustomInspector()).replace('packages://', ''));
```
Comment (arsen2010) [inline editor/inspector/assets/material.js]: need to use 'switch case‘ deel with 'packages://' data

File: editor/inspector/assets/material.js [hunk 5]
```diff
@@ -268,24 +301,25 @@ exports.update = async function(assetList, metaList) {
...
 
     this.updateTechniqueOptions();
     this.setDirtyData();
-
-    // optimize calculate speed when edit multiple materials in node mode
-    requestIdleCallback(() => {
-        this.updatePasses();
-    });
+    if (await this.getCustomInspector()) {
+        this.updateCustomInspector();
```
Comment (arsen2010) [inline editor/inspector/assets/material.js]: if it is in a 'if case' , use function name ’hasCustomInspector‘  would be better

File: editor/inspector/assets/material.js [hunk 6]
```diff
@@ -56,6 +70,25 @@ exports.methods = {
...
 
+    async updateCustomInspector() {
+        this.$.customPanel.hidden = false;
+        this.$.section.hidden = true;
+        this.$.materialDump.hidden = true;
+        try {
+            const relatePath = normalize((await this.getCustomInspector()).replace('packages://', ''));
+            const name = relatePath.split(sep)[0];
+            const packagePath = Editor.Package.getPackages({ name, enable: true })[0].path;
+            const path = join(packagePath, relatePath.split(name)[1]);
```
Comment (arsen2010) [inline editor/inspector/assets/material.js]: before join path, need to judge the package is exist or not.

File: editor/inspector/assets/material.js [hunk 7]
```diff
@@ -80,13 +113,14 @@ exports.methods = {
      * Update the pass data that is finally displayed in the panel
      */
     updatePasses() {
+        this.$.customPanel.hidden = true;
```
Comment (arsen2010) [inline editor/inspector/assets/material.js:126]: if $.customPanel is exist and its ’src‘ is not empty, need to remove attr ’src‘

File: editor/inspector/assets/material.js [hunk 8]
```diff
@@ -56,6 +70,25 @@ exports.methods = {
...
+            const name = relatePath.split(sep)[0];
+            const packagePath = Editor.Package.getPackages({ name, enable: true })[0].path;
+            const path = join(packagePath, relatePath.split(name)[1]);
+            if (this.$.customPanel.getAttribute('src') !== join(path)) {
+                this.$.customPanel.setAttribute('src', path);
+            }
+            this.$.customPanel.update(this.material, this.assetList, this.metaList);
+        } catch (error) {
+            console.error(error);
+            console.error(Editor.I18n.t('ENGINE.assets.material.fail-to-load-custom-inspector', { effect: this.material.effect }));
```
Comment (arsen2010) [inline editor/inspector/assets/material.js:95]: error case should inclued packagePath is not exist.

File: editor/inspector/assets/material.js (hunk 2 above)
Comment (chichinohaha) [reply editor/inspector/assets/material.js:52]: Its return value is now used

File: editor/inspector/assets/material.js (hunk 6 above)
Comment (chichinohaha) [reply editor/inspector/assets/material.js]: There is no need. A try catch is used

File: editor/inspector/assets/material.js (hunk 8 above)
Comment (chichinohaha) [reply editor/inspector/assets/material.js:95]: The error was expected

**Review by arsen2010 — APPROVED** [review]

**Review by VisualSJ — APPROVED** [review]

**Pull request merged**